File: golang/8.binary-tree/main/find_closest_value.go
package main

import (
	"math"

	"github.com/AmirAhmadzadeh/binary_tree/binary_tree"
	"github.com/AmirAhmadzadeh/binary_tree/node"
)

func findClosestValueInBinaryTree(tree *binary_tree.BinaryTree, T int) float64 {
    C := math.Inf(+1) ;
    q := []*node.Node{tree.GetRoot()}


    for len(q) > 0 { 
        v := q[0]; 
        q = q[1:];

        if v.Left() != nil { 
            q = append(q, v.Left()) ;
        }

        if v.Right() != nil { 
            q = append(q, v.Right()) ;
        }

        if math.Abs(C - float64(v.Value()))  < math.Abs(float64(T) - float64(v.Value())) { 
            C = float64(v.Value()); 
        }
    }

    return C; 
}

func findClosestValueInBinaryTreeDfsCall(V *node.Node, T float64, C float64) float64 {
	if V == nil {
		return C
	}
	if math.Abs(T-float64(V.Value())) < math.Abs(T-C) {
		C = float64(V.Value())
	}

	if V.Right() != nil && T > float64(V.Value()) {
		return findClosestValueInBinaryTreeDfsCall(V.Right(), T, C)
	} else if V.Left() != nil && T < float64(V.Value()) {
		return findClosestValueInBinaryTreeDfsCall(V.Left(), T, C)
	}

	return C
}

func findClosestValueInBinaryTreeDfs(binaryTree *binary_tree.BinaryTree, T float64) float64 {
	return findClosestValueInBinaryTreeDfsCall(binaryTree.GetRoot(), T, math.Inf(1))
}

File: ts/8.binary-tree/find-closest-value.ts
import { BinaryTree } from './BinaryTree';
import { Node } from './Node';

export function findClosestNumber(bt: BinaryTree<number>, target: number) {
    let q = [bt.getRoot()] ;
    let dist = Infinity ; 


    while(q.length > 0) { 
        const v = q.shift() as Node<number>; 

        v.left && q.push(v.left) ;
        v.right && q.push(v.right) ;


        if (Math.abs(target - v.value) < Math.abs(target - dist)) { 
            dist = v.value; 
        }
    }

    return dist ;
}

function dfs(
    node: Node<number> | null,
    target: number,
    closest: number
): number {
    if (!node) return closest;

    if (Math.abs(target - node.value) < Math.abs(target - closest)) {
        closest = node.value;
    }

    if (node.right && target > node.value) {
        return dfs(node.right, target, closest);
    } else if (node.left && target < node.value) {
        return dfs(node.left, target, closest);
    } else {
        return closest;
    }
}

export function findClosestNumberDfs(bt: BinaryTree<number>, target: number) {
    return dfs(bt.getRoot(), target, Infinity);
}
